Frameworks/Foundation/NSDictionary.mm

```diff
@@ -793,53 +769,94 @@ - (NSArray*)objectsForKeys:(NSArray*)keys notFoundMarker:(id)notFoundMarker {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return [self keysOfEntriesWithOptions:0 passingTest:predicate];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-- (NSSet*)keysOfEntriesWithOptions:(NSEnumerationOptions)opts passingTest:(BOOL (^)(id, id, BOOL*))predicate {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (NSSet*)keysOfEntriesWithOptions:(NSEnumerationOptions)options passingTest:(BOOL (^)(id, id, BOOL*))predicate {
+    __block NSMutableSet* ret = [NSMutableSet setWithCapacity:[self count]];
+    [self enumerateKeysAndObjectsWithOptions:options
+                                  usingBlock:^(id key, id obj, BOOL* stop) {
+                                      if (predicate(key, obj, stop)) {
+                                          [ret addObject:key];
+                                      }
+                                  }];
+
+    return ret;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes atomically parameter not supported
 */
-- (BOOL)writeToURL:(NSURL*)aURL atomically:(BOOL)flag {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (BOOL)writeToURL:(NSURL*)url atomically:(BOOL)atomically {
+    NSError* error = nil;
+    NSData* data = [NSPropertyListSerialization dataWithPropertyList:self format:NSPropertyListXMLFormat_v1_0 options:0 error:&error];
+    if (error != nil) {
+        return NO;
+    }
+
+    return [data writeToURL:url atomically:atomically];
 }
 
 /**
  @Status Stub
  @Notes
 */
 - (NSString*)descriptionWithLocale:(id)locale {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return [self descriptionWithLocale:locale indent:0];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-- (NSString*)descriptionWithLocale:(id)locale indent:(NSUInteger)level {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (NSString*)descriptionWithLocale:(id)locale indent:(NSUInteger)indent {
+    NSMutableString* s = [NSMutableString new];
+    for (NSUInteger i = 0; i < indent; ++i) {
+        [s appendString:@"    "];
+    }
+
+    [s appendString:@"{\n"];
+
+    {
+        ++indent;
+        auto deferPop = wil::ScopeExit([&indent]() { --indent; });
+
+        // Documentation specifies that output should be sorted by key using -[compare:] for NSString* keys
+        // Order is undefined if keys aren't NSString so just return ascending
+        NSArray* sortedKeys = [[self allKeys] sortedArrayUsingFunction:[](id key1, id key2, void* context) {
+            return ([key1 isKindOfClass:[NSString class]] && [key2 isKindOfClass:[NSString class]]) ? [key1 compare:key2] :
+                                                                                                      NSOrderedAscending;
+        }
+                                                               context:nullptr];
+
+        for (id key in sortedKeys) {
+            for (NSUInteger i = 0; i < indent; ++i) {
+                [s appendString:@"    "];
+            }
```

Comment:
same here, let's use string format to do this.
we end up doing allot of appendStrings :(